third_party/ibis/ibis_addon/operations.py

```diff
@@ -157,201 +125,222 @@ def strftime_bigquery(translator, expr):
 
-def sa_format_hashbytes_mysql(translator, expr):
-    arg, how = expr.op().args
-    compiled_arg = translator.translate(arg)
-    hash_func = sa.func.sha2(compiled_arg, sa.sql.literal_column("'256'"))
+def sa_format_hashbytes_mysql(translator, op):
+    arg = translator.translate(op.arg)
+    hash_func = sa.func.sha2(arg, sa.sql.literal_column("'256'"))
     return hash_func
 
-def sa_format_hashbytes_db2(translator, expr):
-    arg, how = expr.op().args
-    compiled_arg = translator.translate(arg)
+def sa_format_hashbytes_db2(translator, op):
+    compiled_arg = translator.translate(op.arg)
     hashfunc = sa.func.hash(compiled_arg,sa.sql.literal_column("2"))
     hex = sa.func.hex(hashfunc)
     return sa.func.lower(hex)
 
-def sa_format_hashbytes_redshift(translator, expr):
-    arg, how  = expr.op().args
-    compiled_arg = translator.translate(arg)
-    return sa.sql.literal_column(f"sha2({compiled_arg}, 256)")
+def sa_format_hashbytes_redshift(translator, op):
+    arg = translator.translate(op.arg)
+    return sa.sql.literal_column(f"sha2({arg}, 256)")
 
-def sa_format_hashbytes_postgres(translator, expr):
-    arg, how = expr.op().args
-    compiled_arg = translator.translate(arg)
-    convert = sa.func.convert_to(compiled_arg, sa.sql.literal_column("'UTF8'"))
+def sa_format_hashbytes_postgres(translator, op):
+    arg = translator.translate(op.arg)
+    convert = sa.func.convert_to(arg, sa.sql.literal_column("'UTF8'"))
     hash_func = sa.func.sha256(convert)
     return sa.func.encode(hash_func, sa.sql.literal_column("'hex'"))
 
-def sa_format_to_char(translator, expr):
-    arg, fmt = expr.op().args
-    compiled_arg = translator.translate(arg)
-    compiled_fmt = translator.translate(fmt)
-    return sa.func.to_char(compiled_arg, compiled_fmt)
+def sa_format_hashbytes_snowflake(translator, op):
+    arg = translator.translate(op.arg)
+    return sa.func.sha2(arg)
 
-def sa_format_to_stringjoin(translator, expr):
-    sep, elements = expr.op().args
-    return sa.func.concat_ws(translator.translate(sep), *map(translator.translate, elements))
+def sa_format_to_char(translator, op):
+    arg = translator.translate(op.arg)
+    fmt = translator.translate(op.fmt)
+    return sa.func.to_char(arg, fmt)
 
 
-def sa_cast_postgres(t, expr):
-    arg, typ = expr.op().args
+def sa_cast_postgres(t, op):
+    # Add cast from numeric to string
+    arg = op.arg
+    typ = op.to
+    arg_dtype = arg.output_dtype
 
     sa_arg = t.translate(arg)
-    sa_type = t.get_sqla_type(typ)
-
-    # Specialize going from an integer type to a timestamp
-    if isinstance(arg.type(), dt.Integer) and isinstance(sa_type, sa.DateTime):
-        return sa.func.timezone('UTC', sa.func.to_timestamp(sa_arg))
 
     # Specialize going from numeric(p,s>0) to string
-    if isinstance(arg.type(), dt.Decimal) and arg.type().scale > 0 and typ.equals(dt.string):
+    if arg_dtype.is_decimal() and arg_dtype.scale > 0 and typ.equals(dt.string):
```

Comment:
I created a PostgreSQL `decimal` column with no precision or scale and get an error from this line:

```
  File "/usr/local/google/home/neiljohnson/github/professional-services-data-validator/third_party/ibis/ibis_addon/operations.py", line 260, in sa_cast_postgres
    if arg_dtype.is_decimal() and arg_dtype.scale > 0 and typ.equals(dt.string):
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

On the `develop` branch I don't get the exception. But it seems to be unrelated to this line, on develop a scale of `None` becomes `0`. That isn't happening on the feature branch.

Reply:
Looks like the upgrade added the default scale of 'None' which I'll have to handle: https://github.com/ibis-project/ibis/blob/b9977fbe3ed30033fc2500abe061eda8d7e26eed/ibis/expr/datatypes/core.py#L669 

Previously, the scale had to be an int so we didn't get an exception

Reply:
Now I've slept on it I think `None` is correct. This line of code (which I added as it happens) is probably not defensive enough. Perhaps it should be `(arg_dtype.scale or 0) > 0`?

Reply:
I've tested Oracle `NUMBER` vs PostgreSQL `decimal`, both without precision or scale.

On `develop` we get:

```
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 2cae2db8-8dd8-4628-8b40-2db469a9ae1c │ Schema            │ Schema            │ []       │ dvt_test.dvt_ora2pg_types │ dvt_test.dvt_ora2pg_types │ col_num              │ col_num              │ decimal(38, 0)     │ decimal(38, 0)     │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
```

On this feature branch we get:

```
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 80f8ecbc-b090-488e-a089-733766876bcf │ Schema            │ Schema            │ []       │ dvt_test.dvt_ora2pg_types │ dvt_test.dvt_ora2pg_types │ col_num              │ col_num              │ decimal(38, 0)     │ decimal            │ fail                │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
```

In a purists world I think `decimal` with no precision or scale is the correct outcome and therefore the correct action would be to remove the defaults from Oracle. But if this branch should reflect what we had before then perhaps PostgreSQL should have the same (incorrect) defaults as it had on `develop`.

Reply:
I went ahead and removed the defaults from Oracle. I also added the extra condition in this line to `if arg_dtype.scale and arg_dtype.scale > 0` to account for scale being None.

Reply:
I've retested and that works as expected:

```
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ a845a755-eb50-4ad9-ac63-d582a5dc527f │ Schema            │ Schema            │ []       │ dvt_test.dvt_ora2pg_types │ dvt_test.dvt_ora2pg_types │ col_num              │ col_num              │ decimal            │ decimal            │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
```